node/router/src/lib.rs

```diff
@@ -224,49 +205,32 @@ impl<N: Network> Router<N> {
         if self.number_of_connected_peers() >= self.max_connected_peers() {
             bail!("Dropping connection attempt to '{peer_ip}' (maximum peers reached)")
         }
+        // Ensure the node is not already connecting to this peer.
+        if self.is_connecting(&peer_ip) {
+            debug!("Dropping connection attempt to '{peer_ip}' (already connecting)");
+            return Ok(true);
+        }
         // Ensure the node is not already connected to this peer.
         if self.is_connected(&peer_ip) {
             debug!("Dropping connection attempt to '{peer_ip}' (already connected)");
             return Ok(true);
         }
         // Ensure the peer is not restricted.
         if self.is_restricted(&peer_ip) {
-            bail!("Dropping connection attempt to '{peer_ip}' (restricted)")
-        }
-        // Ensure the node is not already connecting to this peer.
-        match self.connecting_peers.lock().entry(peer_ip) {
```

Comment:
I think I added this to prevent certain race conditions. Would it be possible to hold the lock to the peer pool, check the connection attempt, and then insert if successful? 

If I understand correctly, you mark a peer as connecting later when initiating the handshake, but you could already mark it as connecting here.

Reply:
I would love to do it, but it would be a pretty serious attack vector if we optimistically assumed that the peer wouldn't take their time to finish connecting.

As for marking an _outgoing_ connection attempt as "connecting", this is primarily done by the lower-level layer (`tcp`, which also has a notion of "partial" connections), which is sufficient to avoid duplicate attempts. Of course it is trickier in case of inbound connections, but we do it as soon as possible (in `ensure_peer_is_allowed`).

Reply:
Oh, I didn't see the logic in `ensure_peer_is_allowed`. Thanks for clarifying1